client/src/controllers/PreviewController.ts

```diff
@@ -0,0 +1,426 @@
+   */
+  hasChanges() {
+    // https://github.com/microsoft/TypeScript/issues/30584
+    const newPayload = new URLSearchParams(
+      new FormData(this.editForm) as unknown as Record<string, string>,
+    ).toString();
+    const changed = this.formPayload !== newPayload;
+
+    this.formPayload = newPayload;
+    return changed;
+  }
+
+  /**
+   * Initialises the auto update mechanism. This works by comparing the current
+   * form data with the previous form data at a set interval. If the form data
+   * has changed, the preview will be updated. The interval is only active when
+   * the side panel is shown.
+   */
+  initAutoUpdate() {
+    let updateInterval: ReturnType<typeof setInterval>;
+
+    // This controller is encapsulated as a child of the side panel element,
+    // so we need to listen to the show/hide events on the parent element
+    // (the one with [data-side-panel]).
+    // If we had support for data-controller attribute on the side panels,
+    // we could remove the intermediary element and make the [data-side-panel]
+    // element to also act as the controller.
+    const sidePanelContainer = this.element.parentElement as HTMLDivElement;
+
+    // Only set the interval while the panel is shown
+    sidePanelContainer.addEventListener('show', () => {
+      updateInterval = setInterval(
+        this.checkAndUpdatePreview.bind(this),
+        this.autoUpdateIntervalValue,
+      );
+    });
+
+    // Clear the interval when the panel is hidden
+    sidePanelContainer.addEventListener('hide', () => {
+      clearInterval(updateInterval);
+    });
+  }
+
+  /**
+   * Changes the preview mode specified in the iframe and new tab link,
+   * and updates the preview.
+   * @param event Event from the `<select>` element
+   */
+  handlePreviewModeChange(event: Event) {
+    const mode = (event.target as HTMLSelectElement).value;
+    const url = new URL(this.baseUrlValue, document.location.href);
+
+    // Update the new tab link
+    url.searchParams.set('mode', mode);
+    this.newTabButtonTarget.href = url.toString();
+
+    // Make sure data is updated and an alert is displayed if an error occurs
+    this.handlePreview();
+  }
+
+  connect() {
+    this.observePanelSize();
+
+    this.editForm = document.querySelector<HTMLFormElement>(
+      '[data-edit-form]',
+    ) as HTMLFormElement;
+
+    if (!this.baseUrlValue) {
+      throw new Error(
+        `The preview panel controller requires the data-${this.identifier}-base-url-value attribute to be set`,
+      );
+    }
+
+    if (this.autoUpdateValue) {
+      this.initAutoUpdate();
+    }
+
+    this.restoreLastSavedPreferences();
+  }
+
+  /**
+   * Restore the last saved preferences.
+   * Currently, only the last selected device size is restored.
+   */
+  restoreLastSavedPreferences() {
+    // Remember last selected device size
+    let lastDevice: string | null = null;
+    try {
+      lastDevice = localStorage.getItem('wagtail:preview-panel-device');
```

Comment:
Might be nice to have this as a stimulus value that's just defaulted, no major difference to code but allows for easy reuse / sharing of controller code.